Add option to set name of the local cluster_node

cneumann opened the pull request:
Before the cluster_node config element that applies to the local
machine is determined by comparing the machine's hostname with
the host_name property on the config element.
That makes it very difficult to run a cluster on the local machine for
testing purposes. If mLocalNodeName is set (is non-empty) its value is
used to determine the config element that applies to the local machine,
otherwise the existing behavior using the hostname is used.

If this is considered useful I can write a followup patch to make the kernel recognize a --vrj_node_name option (in addition to --vrjmaster,--vrjslave,--listen_port) that passes its value to the ClusterManager. Currently I'm doing this for my applications as part of the application's command line processing.

Commit:
Add option to set name of the local cluster_node

Before the cluster_node config element that applies to the local
machine is determined by comparing the machine's hostname with
the host_name property on the config element.
That makes it very difficult to run a cluster on the local machine for
testing purposes. If mLocalNodeName is set (is non-empty) its value is
used to determine the config element that applies to the local machine,
otherwise the existing behavior using the hostname is used.

tfurlong commented:
Carsten, can you post how you use this code in your application?  I'm all for this feature, but I wonder if we could accomplish a similar goal by setting an environment variable (call it VJ_OVERRIDE_HOSTNAME) before kicking off each executable instance instead of altering applications to allow this sort of testing.

cneumann commented:
My application class has a member function to process the command line arguments:
MyApp::processCommandline(int argc, char* argv[])

where I process vrjuggler and application specific command line options with boost::program_options:

    desc.add_options()
        ("node-name", bpo::value<std::string>(&nodeName_),
        "name of the local node in the .jconf files");

    // add VR Juggler options
    desc.add(mKernel->getGeneralOptions());
    desc.add(mKernel->getClusterOptions());
    desc.add(mKernel->getConfigOptions());

    // pass the node name to ClusterManager
    if(vm.count("node-name"))
    {
        cluster::ClusterManager* clusterMgr{cluster::ClusterManager::instance()};
        clusterMgr->setLocalNodeName(nodeName_);
    }

    mKernel->init(vm);

The client nodes are passed the following command line arguments from a script that sets up LD_LIBRARY_PATH/PATH, VJ_BASE_DIR, etc.:

app --vrjslave --listen_port=54321 --node-name=ClusterNode-Center

where my config files contain an element: `<cluster_node name="ClusterNode-Center" version="1">` matching the name.

I'm not a huge fan of environment variables as the only means of controlling a feature, but how about this: if the VJ_OVERRIDE_HOSTNAME env variable is set its value is used as the value of ClusterManager::mLocalNodeName?

tfurlong commented:
Carsten, can we make this feature transparent to the app developer?  Meaning, the command-line would override the default method of obtaining the hostname?  I think this is a good core feature to have.  I'm OK with no environment variable.  I haven't had the time to put the attention to this that I'd like to, so maybe I'm missing a reason why we can't do that.

cneumann commented:
Do you mean add (modulo different name) a command line option --vrjnodename that is recognized by vrj::Kernel (just like --vrjslave, --listen_port, etc)? That's certainly possible. I'll see if I can get around to implementing it this week; I'll update the PR when I have something. Thanks!

tfurlong commented:
Yes, exactly, thanks!

Commit:
Tweak comment, variable names for setLocalNodeName

Commit:
Add --vrjnodename command line option

If given this sets ClusterManager::setLocalNodeName to the option value.
In that case ClusterManager will recognize the config element for
the local node by comparing the config element's name with the option
value. Otherwise the existing behavior of using the host name is used.

cneumann commented:
Sorry, this took longer than I had hoped :-/
I've updated the branch with the changes adding the command line option (called --vrjnodename).

tfurlong commented:
Looks great!  Thanks Carsten.
